files/private-chef-ctl-commands/key-control.rb

```diff
@@ -14,345 +14,425 @@
+    exit 1
+  end
 
+  def parse_invalid_arg_error(err)
+    arg = err.message.match(/(?<=invalid option:\s).*/)[0]
+    STDERR.puts "Error: Invalid argument #{arg} detected. Please remove or use a different command."
+    exit 1
+  end
 
-    if expiration_date
-      begin
-        Time.parse(expiration_date)
-      rescue
-        STDERR.puts "Error: Invalid date format inserted"
-        STDERR.puts "Error: You inserted #{expiration_date}"
-        STDERR.puts "Error: Please input a date in format YYYY-MM-DDTHH:MM:SS e.g. 2013-12-24T21:00:00"
-        STDERR.puts "Error: Alternatively, simply don't pass --expiration-date if you don't want your key to expire"
-        exit 1
-      end
-    else
-      expiration_date = "infinity"
+  def exit_http_fail
+    STDERR.puts "Error: An unexpected error has occured (the server returned a #{e.response.code})."
+    STDERR.puts "Error: Please contact a system admin if the problem persists."
+    exit 1
+  end
+
+  def check_field_exists(options, usage, field_symbol, arg_flag)
+    if options[field_symbol].nil?
+      STDERR.puts usage
+      STDERR.puts "Error: Please pass a valid value to #{arg_flag}."
+      exit 1
     end
+  end
 
-    object = {
-      'id'          => id,
-      'key_name'    => key_name,
-      'public_key'  => key,
-      'key_version' => 0,
-      'created_at'  => Time.now.utc,
-      'expires_at'  => expiration_date
+  def build_key_object(name, key, expiration_date)
+    {
+      'name' => name,
+      'public_key' => key,
+      'expiration_date' => expiration_date
     }
+  end
 
-    begin
-      database.db[:keys].insert(object)
-    rescue Sequel::UniqueConstraintViolation
-      STDERR.puts "Error: The key_name you passed (#{key_name}) was not unique for your object (#{objectname})"
-      exit 1
+  def output_key_results(results, show_public_keys, rest)
+    results.each do |result|
+      puts "\nname: #{result['name']}"
+      puts "expired: #{result['expired']}"
+      if show_public_keys
+        public_key = rest.get_rest(result["uri"])
+        puts "public_key:"
+        puts public_key["public_key"]
+      end
     end
   end
 end
 
 add_command_under_category "add-client-key", "key-rotation", "Create a new client key", 2 do
-  database = Database.new(running_config)
+  @options = OpenStruct.new
+  @options.expiration_date = "infinity"
+  @key = nil
+  @helper = Helper.new
+  @usage = "Usage: chef-server-ctl add-client-key -o, --orgname <name> -c, --clientname <name> -p, --public-key-path <path_to_public_key> [-e, --expiration-date <date>] [-k, --key-name <name>]"
+  @usage = @usage << @helper.add_key_usage
+  @arg_list = ["-o", "--orgname", "-c", "--clientname", "-p", "--public-key-path", "-e", "--expiration-date", "-k", "--key-name"]
+
+  opt_parser = OptionParser.new do |opts|
+    opts.banner = @usage
+
+    opts.on("-o", "--orgname <name>", "Name of the organization that your client is in.") do |orgname|
```

Comment:
Ah you are correct. Updating to take advantage of this. I read the optparse docs but they weren't super helpful.